File: app/actions.py
import psycopg2
from psycopg2.extras import RealDictCursor

# Database connection parameter configuration
CONN_STRING = "dbname=lab_db user=postgres password=postgres host=127.0.0.1 port=5432"


def get_connection():
    return psycopg2.connect(CONN_STRING)


def save_user(first_name: str, last_name: str, phone: str, id_card: str) -> int:

    sql = "SELECT save_users(%s, %s, %s, %s);"
    
    with get_connection() as conn:
        with conn.cursor() as cursor:
            cursor.execute(sql, (first_name, last_name, phone, id_card))
            result = cursor.fetchone()
            conn.commit()
            return result[0] if result else None


def check_exists(id_card: str) -> bool:

    sql = "SELECT COUNT(1) FROM users WHERE id_card_num = %s;"
    
    with get_connection() as conn:
        with conn.cursor() as cursor:
            cursor.execute(sql, (id_card,))
            count = cursor.fetchone()[0]
            return count > 0


def delete_tier(tier_id: int) -> tuple[bool, str]:

    sql = "SELECT delete_tiers(%s);"
    
    try:
        with get_connection() as conn:
            with conn.cursor() as cursor:
                cursor.execute(sql, (tier_id,))
                conn.commit()
                return True, ""
    except psycopg2.Error as ex:
        # Captures RAISE EXCEPTION messages thrown by PL/pgSQL
        return False, ex.pgerror or str(ex)


def filter_users(search_text: str) -> list[tuple]:

    sql = """
        SELECT user_id, first_name, last_name, phone_number, id_card_num
        FROM users
        WHERE LOWER(first_name) LIKE LOWER(%s)
           OR LOWER(last_name) LIKE LOWER(%s)
           OR LOWER(id_card_num) LIKE LOWER(%s);
    """
    search_pattern = f"%{search_text}%"
    
    with get_connection() as conn:
        with conn.cursor() as cursor:
            cursor.execute(sql, (search_pattern, search_pattern, search_pattern))
            return cursor.fetchall()


def get_parent_list() -> list[tuple[int, str]]:
    sql = "SELECT user_id, first_name || ' ' || last_name || ' (' || id_card_num || ')' FROM users;"
    
    with get_connection() as conn:
        with conn.cursor() as cursor:
            cursor.execute(sql)
            return cursor.fetchall()

File: app/csharplab/actions.cs
using System;
using System.Collections.Generic;
using Npgsql;

public class DatabaseRepository
{
    private readonly string _connectionString;

    public DatabaseRepository(string connectionString)
    {
        _connectionString = connectionString;
    }

    // 1. Loads parent table records for foreign key dropdown selectors
    public List<KeyValuePair<long, string>> GetParentList()
    {
        var result = new List<KeyValuePair<long, string>>();

        using (var conn = new NpgsqlConnection(_connectionString))
        {
            conn.Open();
            string sql = "SELECT user_id, first_name || ' ' || last_name || ' (' || id_card_num || ')' FROM users";

            using (var cmd = new NpgsqlCommand(sql, conn))
            using (var reader = cmd.ExecuteReader())
            {
                while (reader.Read())
                {
                    result.Add(new KeyValuePair<long, string>(
                        reader.GetInt64(0),
                        reader.GetString(1)
                    ));
                }
            }
        }

        return result;
    }

    // 2. Case-insensitive filtering via SQL query
    public List<string[]> FilterRecords(string searchText)
    {
        var records = new List<string[]>();

        using (var conn = new NpgsqlConnection(_connectionString))
        {
            conn.Open();
            string sql = @"SELECT user_id, first_name, last_name, phone_number, id_card_num 
                           FROM users 
                           WHERE LOWER(first_name) LIKE LOWER(@search) 
                              OR LOWER(last_name) LIKE LOWER(@search)
                              OR LOWER(id_card_num) LIKE LOWER(@search)";

            using (var cmd = new NpgsqlCommand(sql, conn))
            {
                cmd.Parameters.AddWithValue("search", $"%{searchText}%");

                using (var reader = cmd.ExecuteReader())
                {
                    while (reader.Read())
                    {
                        records.Add(new string[]
                        {
                            reader.GetInt64(0).ToString(),
                            reader.GetString(1),
                            reader.GetString(2),
                            reader.GetString(3),
                            reader.GetString(4)
                        });
                    }
                }
            }
        }

        return records;
    }

    // 3. Checks whether a record already exists before saving
    public bool CheckExists(string idCardNum)
    {
        using (var conn = new NpgsqlConnection(_connectionString))
        {
            conn.Open();
            string sql = "SELECT COUNT(1) FROM users WHERE id_card_num = @idc";

            using (var cmd = new NpgsqlCommand(sql, conn))
            {
                cmd.Parameters.AddWithValue("idc", idCardNum);
                long count = Convert.ToInt64(cmd.ExecuteScalar());
                return count > 0;
            }
        }
    }

    // 4. Calls PL/pgSQL function to insert or update user records
    public long SaveRecord(string firstName, string lastName, string phoneNumber, string idCardNum)
    {
        using (var conn = new NpgsqlConnection(_connectionString))
        {
            conn.Open();
            using (var cmd = new NpgsqlCommand("SELECT save_users(@fn, @ln, @pn, @idc)", conn))
            {
                cmd.Parameters.AddWithValue("fn", firstName);
                cmd.Parameters.AddWithValue("ln", lastName);
                cmd.Parameters.AddWithValue("pn", phoneNumber);
                cmd.Parameters.AddWithValue("idc", idCardNum);

                return Convert.ToInt64(cmd.ExecuteScalar());
            }
        }
    }

    // 5. Calls PL/pgSQL function to delete tier records with constraint error handling
    public bool DeleteRecord(int tierId, out string errorMessage)
    {
        errorMessage = string.Empty;

        try
        {
            using (var conn = new NpgsqlConnection(_connectionString))
            {
                conn.Open();
                using (var cmd = new NpgsqlCommand("SELECT delete_tiers(@id)", conn))
                {
                    cmd.Parameters.AddWithValue("id", tierId);
                    cmd.ExecuteNonQuery();
                    return true;
                }
            }
        }
        catch (PostgresException ex)
        {
            errorMessage = ex.MessageText;
            return false;
        }
    }
}